full-service/src/db/txo.rs

```diff
@@ -537,146 +436,163 @@ impl TxoModel for Txo {
             .load(conn)?;
+        Ok(results)
+    }
 
-        let details: Result<Vec<TxoDetails>, WalletDbError> = results
-            .iter()
-            .map(|t| Txo::get(&t.txo_id_hex, &conn))
-            .collect();
-        details
+    fn list_unspent(
+        account_id_hex: &str,
+        assigned_subaddress_b58: Option<&str>,
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<Vec<Txo>, WalletDbError> {
+        use crate::db::schema::txos;
+
+        let results = txos::table
+            .filter(txos::received_account_id_hex.eq(account_id_hex))
+            .filter(txos::subaddress_index.is_not_null())
+            .filter(txos::pending_tombstone_block_index.is_null())
+            .filter(txos::spent_block_index.is_null());
+
+        let txos: Vec<Txo> = if let Some(subaddress_b58) = assigned_subaddress_b58 {
+            let subaddress = AssignedSubaddress::get(&subaddress_b58, conn)?;
+            results
+                .filter(txos::subaddress_index.eq(subaddress.subaddress_index))
+                .load(conn)?
+        } else {
+            results.load(conn)?
+        };
+
+        Ok(txos)
     }
 
-    fn list_by_status(
+    fn list_spent(
         account_id_hex: &str,
-        status: &str,
+        assigned_subaddress_b58: Option<&str>,
         conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
     ) -> Result<Vec<Txo>, WalletDbError> {
-        use crate::db::schema::{account_txo_statuses, txos};
-
-        let results: Vec<Txo> = txos::table
-            .inner_join(
-                account_txo_statuses::table.on(txos::txo_id_hex
-                    .eq(account_txo_statuses::txo_id_hex)
-                    .and(account_txo_statuses::account_id_hex.eq(account_id_hex))
-                    .and(account_txo_statuses::txo_status.eq(status))),
-            )
-            .select(txos::all_columns)
+        use crate::db::schema::txos;
+
+        let results = txos::table
+            .filter(txos::received_account_id_hex.eq(account_id_hex))
+            .filter(txos::subaddress_index.is_not_null())
+            .filter(txos::pending_tombstone_block_index.is_null())
+            .filter(txos::spent_block_index.is_not_null());
+
+        let txos: Vec<Txo> = if let Some(subaddress_b58) = assigned_subaddress_b58 {
+            let subaddress = AssignedSubaddress::get(&subaddress_b58, conn)?;
+            results
+                .filter(txos::subaddress_index.eq(subaddress.subaddress_index))
+                .load(conn)?
+        } else {
+            results.load(conn)?
+        };
+
+        Ok(txos)
+    }
+
+    fn list_secreted(
+        account_id_hex: &str,
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<Vec<Txo>, WalletDbError> {
+        use crate::db::schema::txos;
+
+        let txos: Vec<Txo> = txos::table
+            .filter(txos::received_account_id_hex.ne(account_id_hex))
```

Comment:
From someone that understands very little of the full-service internals: Wouldn't this return all txos that were received by other accounts? what is the purpose of this? what does `secreted` means?

Reply:
Secreted means minted by this account, but not received by this account. This query was in fact wrong, good catch!